src/main/java/com/google/research/bleth/simulator/Agent.java

```diff
@@ -0,0 +1,11 @@
+package com.google.research.bleth.simulator;
+
+/** Represent both observer and beacon. */
+public interface Agent {
+
+    /**
+     * Calculate the location the agent is moving to, based on its current location and its strategy.
+     * @return the location on board which the agent is moving to.
+     */
+    Location move();
```

Comment:
It appears each Agent is aware of its own location; how do you ensure the Agent and any associated Board are truly in sync about this? (I expect we want to keep an invariant of "for every Board that has a reference to Agent _a_, _a_'s location on the board is the same as its internal notion of location"? Or is this sometimes broken, e.g. when the Board is an estimated one?)

Reply:
Both the real and estimated boards are instances of the Board class. However, this invariant should be kept w.r.t the real board only. It works as follows:

- Each agent is aware of its location on the real board.
- The agent's move method updates the agent's internal notion of location and returns this new location.
- The real board updates its matrix by calling moveAgent.

The resolver however doesn't use the agent's method move at all, it updates its board using the estimations based on the transmissions it receives.
The agent isn't aware at all to its location on the estimated board.

Reply:
I still don't like this separation. It looks too fragile to me (as in: we have no guarantee that *only* the real Board will invoke Agent.move(), or we break an invariant - and we have no way to even detect that the invariant was broken).

Let's discuss alternatives?

Reply:
Option 1:
Board will be an abstract class, we'll also have RealBoard and EstimatedBoard subclasses. The placeAgent() will be the only abstract method of Board. So now we have:

- Each agent is aware of its location on the real board.
- In addition, the Agent have a setRealLocation public method.
- The agent's move method returns this new location but does not update the agent's internal notion of location.
- The real board updates its matrix by calling placeAgent(someAgent.move(), someAgent).
- Only for RealBoard: Inside the placeAgent method the setRealLocation is being called to update the agent's internal notion of location.

This way both internal and global location on the real board are being updated in the same method.

Option 2:
The simulator itself will be responsible for both updating the real board and the agent's internal notion of location (using the board's placeAgent method and the agent's setter for location).